feat: implement Operator Fee Fix (Jovian)

fakedev9999 opened the pull request:
This is a consensus breaking change.

**Description**

Updates operator fee formula to be more useful.

🤝 Corresponding op-geth PR: https://github.com/ethereum-optimism/op-geth/pull/676

**Tests**

Added new tests for updated operator fee formula on Jovian.

**Additional context**

- #16623 
- https://github.com/ethereum-optimism/design-docs/pull/299

**Metadata**

- Closes #17002.

geoknee commented:
Do you think we need to make any updates here?

https://github.com/ethereum-optimism/optimism/blob/636ad830f50b859366f9e15654c34d5f875c83a8/op-node/rollup/derive/system_config.go#L162-L169

It seems like perhaps not, but wanted to get your take. Somewhere we need to have sufficient test coverage to give confidence about the behaviour pre/post Jovian with the new and old data types.

geoknee commented:
/ci authorize [c122745](https://github.com/ethereum-optimism/optimism/pull/17366/commits/c122745b1346287b3e1c1dc6839b71aa4252178d)

fakedev9999 committed:
fix!: multiply operatorFeeScalar by 100 instead of dividing by 1e6

fakedev9999 commented:
Hey @geoknee, sorry for the confusion, initially I misunderstood the scope of the fix and changed the type to uint64. The new fix I force pushed **only changes the multiplier** from 1e-6 to 100.

Since the operatorFeeScalar remains uint32, so I do not think we need changes to `system_config.go`.

fakedev9999 committed:
chore: bump to 1.5.0

fakedev9999 committed:
chore: bump to 2.0.0

geoknee commented:
General comment here is that we need to be adding _new_ code and tests for the behaviour, because we continue to use the existing Isthmus formula up until the Jovian fork.

fakedev9999 committed:
refac: gate at Jovian

fakedev9999 committed:
missing changes

fakedev9999 committed:
update

fakedev9999 committed:
reduce diff

fakedev9999 committed:
snapshot abi + storage layout + remove unncessary change

geoknee commented on packages/contracts-bedrock/src/L2/GasPriceOracle.sol:
```diff
@@ -194,15 +208,35 @@ contract GasPriceOracle is ISemver {
+    ///      - Pre-Isthmus: Returns 0 (no operator fee)
+    ///      - Isthmus (pre-Jovian): operatorFee = (gasUsed * operatorFeeScalar / 1e6) + operatorFeeConstant
+    ///      - Jovian and after: operatorFee = (gasUsed * operatorFeeScalar * 100) + operatorFeeConstant
+    ///      The Jovian formula provides better precision for small scalar values.
```
```suggestion
```
I don't think we need this here.

geoknee commented on packages/contracts-bedrock/src/L2/GasPriceOracle.sol:
```diff
@@ -194,15 +208,35 @@ contract GasPriceOracle is ISemver {
+                constant_
+            );
+        } else {
+            // Original Isthmus formula: divide by 1e6
```
```suggestion
            return Arithmetic.saturatingAdd(
                Arithmetic.saturatingMul(_gasUsed, scalar) * 100,
                constant_
            );
        } else {
```
I think the code is sufficiently self documenting, here.

geoknee commented on op-devstack/dsl/bridge.go:
```diff
@@ -571,9 +571,11 @@ func gasCost(rcpt *types.Receipt) eth.ETH {
 		// Also see: https://specs.optimism.io/protocol/isthmus/exec-engine.html#operator-operatorCost
+		// TODO: This currently uses the Jovian formula (multiply by 100). Need to update to handle both
+		// Isthmus (divide by 1e6) and Jovian formulas based on fork activation.
 		operatorCost := new(big.Int).SetUint64(rcpt.GasUsed)
```
I would suggest we need to thread some more information to this helper function. The outer test context should be able to pass the chain config and block time, or some interface which can be used to infer which fork is active. So the function signature needs to be expanded with this information

fakedev9999 replied:
@geoknee Thanks! Got the fixes in :)

geoknee commented:
/ci authorize [134edb3](https://github.com/ethereum-optimism/optimism/pull/17366/commits/134edb32991f408a3354caf3de8a68145437cfc9)

fakedev9999 committed:
apply comments

geoknee commented:
/ci authorize [db1dd9a](https://github.com/ethereum-optimism/optimism/pull/17366/commits/db1dd9a9bfb29100740dedfeedee620c6ee0dc23)

fakedev9999 committed:
make ci green

geoknee commented:
/ci authorize [de1eb1c](https://github.com/ethereum-optimism/optimism/pull/17366/commits/de1eb1ca2e4ee0b244b0e41e1bfc99708bb3b53e)

tynes commented on packages/contracts-bedrock/src/L2/GasPriceOracle.sol:
```diff
@@ -194,15 +208,26 @@ contract GasPriceOracle is ISemver {
+    ///      - Isthmus (pre-Jovian): operatorFee = (gasUsed * operatorFeeScalar / 1e6) + operatorFeeConstant
+    ///      - Jovian and after: operatorFee = (gasUsed * operatorFeeScalar * 100) + operatorFeeConstant
+    /// @param _gasUsed The amount of gas used by the transaction
+    /// @return The calculated operator fee
```
This looks correct to me

fakedev9999 committed:
chore: update go.mod

geoknee commented:
/ci authorize [c2704f2](https://github.com/ethereum-optimism/optimism/pull/17366/commits/c2704f2c7c902d9823d34912a5aca29fefeb7b18)

tynes commented on go.sum:
```diff
@@ -230,14 +230,14 @@ github.com/emicklei/dot v1.6.2 h1:08GN+DD79cy/tzN6uLCT84+2Wk9u+wvqP+Hkx/dIR8A=
 github.com/ethereum/c-kzg-4844/v2 v2.1.0/go.mod h1:TC48kOKjJKPbN7C++qIgt0TJzZ70QznYR7Ob+WXl57E=
 github.com/ethereum/go-verkle v0.2.2 h1:I2W0WjnrFUIzzVPwm8ykY+7pL2d4VhlsePn4j7cnFk8=
 github.com/ethereum/go-verkle v0.2.2/go.mod h1:M3b90YRnzqKyyzBEWJGqj8Qff4IDeXnzFw0P9bFw3uk=
+github.com/fakedev9999/op-geth v0.0.0-20250925030841-befc452df542 h1:JWZDNK3TfyZiUBGXX7YiHWmMkosttbyulumc97zfLr4=
```
We are going to need the correct dependency here

tynes replied:
Bump on this

geoknee replied:
I don't think we can update this until the op-geth PR is merged, because it comes from an external PR. We still need to do so before merging this PR but it need not block the review / approval.

leruaa committed:
fix: activate Jovian in e2e tests

leruaa committed:
feat: add Jovian upgrate transactions

leruaa committed:
fix: update GPO bytecode

geoknee commented:
/ci authorize [e3a014d](https://github.com/ethereum-optimism/optimism/pull/17366/commits/e3a014d08726c2a9a6d0e0177cba4ffa3d44b5e6)

geoknee commented on op-devstack/dsl/operator_fee.go:
```diff
@@ -140,8 +147,18 @@ func (of *OperatorFee) ValidateTransactionFees(from *EOA, to *EOA, amount *big.I
+
 		operatorFee := new(big.Int).Mul(big.NewInt(int64(receipt.GasUsed)), big.NewInt(int64(expectedScalar)))
-		operatorFee.Div(operatorFee, big.NewInt(1000000))
+		if isJovian {
```
I was expecting to see a new acceptance test in optimism/op-acceptance-tests/tests/jovian. I think we can't just rely on existing Isthmus acceptance tests because they are not guaranteed to run on a Jovian devnet.

tynes replied:
Perhaps this can be modularized in a way where the same code is imported and ran in both

geoknee replied:
Did that here [21f1632](https://github.com/ethereum-optimism/optimism/pull/17366/commits/21f1632ce498c94315fe703f7b54c8ae503736bf)

geoknee commented on op-devstack/dsl/bridge.go:
```diff
@@ -563,18 +569,47 @@ func (w *Withdrawal) WaitForDisputeGameResolved() {
 		operatorCost.Mul(operatorCost, new(big.Int).SetUint64(*rcpt.OperatorFeeScalar))
-		operatorCost = operatorCost.Div(operatorCost, big.NewInt(1_000_000))
-		operatorCost = operatorCost.Add(operatorCost, new(big.Int).SetUint64(*rcpt.OperatorFeeConstant))
+		if rollupCfg.IsJovian(*blockTimestamp) {
```
Can we use the `IsOperatorFeeFix` flag here?

leruaa replied:
Done

fakedev9999 committed:
use IsOperatorFeeFix

fakedev9999 committed:
add jovian operator fee acceptance test

tynes commented on op-node/rollup/derive/jovian_upgrade_transactions.go:
```diff
@@ -0,0 +1,86 @@
+	GasPriceOracleJovianDeployerAddress = common.HexToAddress("0x4210000000000000000000000000000000000006")
+	jovianGasPriceOracleAddress         = crypto.CreateAddress(GasPriceOracleJovianDeployerAddress, 0)
+
+	// Bytecodes
```
Do you mind including the commit hash that you pulled this from? Just so its possible to recreate easily as part of the review process

leruaa replied:
Done

tynes commented on op-service/txinclude/isthmus_cost_oracle.go:
```diff
@@ -17,13 +17,15 @@ import (
 
-// IsthmusCostOracle implements OPCostOracle only for the Isthmus hard fork.
+// IsthmusCostOracle implements OPCostOracle for the Isthmus and Jovian hard forks.
 type IsthmusCostOracle struct {
```
Why did you decide on adding an `isJovian` flag rather than add a new struct?

fakedev9999 replied:
@tynes It was to avoid code duplication.

I also considered creating a new JovianCostOracle struct, but since the only difference is the multiplier, it felt like overkill to me.

If the logic diverges more substantially in future hard forks, refactoring to separate structs would definitely make sense.

Let me know what you think.

tynes replied:
Semantically its a bit weird but I see the tradeoff. I am fine with this for now to reduce diff

geoknee replied:
I came up with a compromise here [8b3d4a1](https://github.com/ethereum-optimism/optimism/pull/17366/commits/8b3d4a18192d325d10cc2cf1d20807eb464a86a3) which doesn't duplicate code but just names things in a way that's less jarring.

tynes commented on op-service/txinclude/txbudget.go:
```diff
@@ -97,6 +97,8 @@ func (b *TxBudget) AfterIncluded(budgetedCost eth.ETH, tx *IncludedTx) {
 	if receipt.OperatorFeeScalar != nil {
 		// https://github.com/ethereum-optimism/op-geth/blob/6005dd53e1b50fe5a3f59764e3e2056a639eff2f/core/types/rollup_cost.go#L244-L247
 		// Also see: https://specs.optimism.io/protocol/isthmus/exec-engine.html#operator-operatorCost
+		// TODO: This currently uses the Isthmus formula (divide by 1e6). Need to update to handle both
```
Just commenting on an open TODO

geoknee replied:
@joshklop do you have suggestions here? I would like to know if it is OK to do this in a followup -- seems like it is only testing code that uses this and the impact would be fairly minimal / zero not updating it now. And getting it fully correct will require quite a lot of changes.

joshklop replied:
Yes, can be done in a follow-up. It would be good to mark this TODO with an open issue for tracking purposes. Thanks for the callout.

joshklop replied:
[Here](https://github.com/ethereum-optimism/optimism/issues/17817) is the ticket. Let's mark the todo like `TODO(17817): ...`

geoknee replied:
[165c19b](https://github.com/ethereum-optimism/optimism/pull/17366/commits/165c19b292810ae459c77995ac118bc8724dce65)

tynes commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/17366/commits/b78cbfae5129c5e8a1bb4dd74f0718520026d26f

leruaa committed:
chore: add commit hash for GPO bytecode

tynes commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/17366/commits/11000a008689a08856e38e1556160e8ad8f6b974

tynes commented:
It looks like this needs a rebase for dependency conflicts

leruaa committed:
chore: remove useless replace in go.mod

leruaa commented:
@tynes rebased

sebastianst reviewed (commented):
_Sorry, I forgot we choose a different design, only updating the formula._

geoknee commented:
/ci authorize [edaa994](https://github.com/ethereum-optimism/optimism/pull/17366/commits/edaa994761bcb1cc69d655a6e06d808bb33ad977)

geoknee committed:
acceptance-tests: push common code for operator fee into dsl

Also use block timestamp to infer isJovian and assert consistency in GPO.

geoknee committed:
tweak

geoknee committed:
refactor IsthmusCostOracle

geoknee committed:
fix link

geoknee commented:
/ci authorize [8b3d4a1](https://github.com/ethereum-optimism/optimism/pull/17366/commits/8b3d4a18192d325d10cc2cf1d20807eb464a86a3)

geoknee committed:
tag TODO for followup

geoknee committed:
update op-geth

geoknee commented:
/ci authorize [dd390af](https://github.com/ethereum-optimism/optimism/pull/17366/commits/dd390afd93d86fee696e18a17970f035490da335)

geoknee committed:
lint

geoknee commented:
/ci authorize [13821d1](https://github.com/ethereum-optimism/optimism/pull/17366/commits/13821d1462e682b5f01713b9d9155fce9b42a313)

geoknee committed:
lint

geoknee committed:
revert changes to op-service/txinclude

According to @joshklop we don't need to make any changes here at the current time, the code is not actually being used anywhere apart from some skipped tests.

geoknee commented:
/ci authorize [4a85283](https://github.com/ethereum-optimism/optimism/pull/17366/commits/4a852838a89e3068481687a37ebc4aba94be732e)

geoknee committed:
Merge remote-tracking branch 'origin/develop' into taehoon/update-operator-fee-formula

geoknee committed:
Merge remote-tracking branch 'origin/develop' into taehoon/update-operator-fee-formula

geoknee commented:
/ci authorize [b3299e0](https://github.com/ethereum-optimism/optimism/pull/17366/commits/b3299e0ddb55442e6496512084d16c439ea2da77)

geoknee committed:
regenerate l1block bytecode and resolve nonce clash

geoknee commented:
/ci authorize [82a7658](https://github.com/ethereum-optimism/optimism/pull/17366/commits/82a76589e48125519289fa216ecee83b2c04b369)

geoknee committed:
decouple upgrade transactions, change API to getter and fix tests

geoknee committed:
update op-geth reference

geoknee commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -154,11 +154,13 @@ contract L2Genesis is Script {
         }
 
-        if (forkEquals(_fork, Fork.INTEROP)) {
+        activateJovian();
+
+        if (forkEquals(_fork, Fork.JOVIAN)) {
             return;
         }
```
dismiss, this script activates forks in the GPO in order and returns _after_ calling the `activateFork()` helper if that fork is the selected input fork. Since interop moved after Jovian, this change looks correct.

sebastianst commented on op-node/rollup/derive/jovian_upgrade_transactions.go:
```diff
@@ -15,11 +15,25 @@ var (
 	deployJovianL1BlockSource      = UpgradeDepositSource{Intent: "Jovian: L1 Block Deployment"}
 	updateJovianL1BlockProxySource = UpgradeDepositSource{Intent: "Jovian: L1 Block Proxy Update"}
-	L1BlockJovianDeployerAddress   = common.HexToAddress("0x4210000000000000000000000000000000000006")
+	L1BlockJovianDeployerAddress   = common.HexToAddress("0x4210000000000000000000000000000000000007")
```
we may want to either reorder the transaction deployer addresses or the upgrade transactions, see https://github.com/ethereum-optimism/specs/pull/799#pullrequestreview-3340308418

geoknee replied:
I can handle this one.

geoknee replied:
[1fbc84c](https://github.com/ethereum-optimism/optimism/pull/17366/commits/1fbc84c9c07ca012e43dc1d0f632c65174627de3)

geoknee replied:
[1fbc84c](https://github.com/ethereum-optimism/optimism/pull/17366/commits/1fbc84c9c07ca012e43dc1d0f632c65174627de3)

sebastianst reviewed (approved):
lgtm besides comment about reordering upgrade tx or deployer addresses

tynes commented on go.mod:
```diff
@@ -308,7 +308,7 @@ require (
 )
 
-replace github.com/ethereum/go-ethereum => github.com/ethereum-optimism/op-geth v1.101603.2-0.20251014161648-ded9c88e417b
+replace github.com/ethereum/go-ethereum => github.com/fakedev9999/op-geth v0.0.0-20251015121023-b26d533c7fd6
```
We cannot merge this until this correctly points to the `ethereum-optimism` repo

geoknee replied:
The original plan was to:
* wait for this PR to be approved
* merge the op-geth PR
* update this PR to point to the `ethereum-optimism` repo, with a commit from the `optimism` (default) branch
* then merge this PR

geoknee replied:
[866499f](https://github.com/ethereum-optimism/optimism/pull/17366/commits/866499f22e73ec9c8c94cb63a9490da305e5a128)

tynes reviewed (changes requested):
Dependencies are incorrect, must use `ethereum-optimism/op-geth`, not a fork

mds1 commented on packages/contracts-bedrock/src/L2/GasPriceOracle.sol:
```diff
@@ -194,15 +208,29 @@ contract GasPriceOracle is ISemver {
+
+        if (isJovian) {
+            // New formula (post-Jovian): multiply by 100
+            return Arithmetic.saturatingAdd(Arithmetic.saturatingMul(_gasUsed, operatorScalar) * 100, operatorConstant);
```
@fakedev9999 this comment was resolved, but it seems correct to me (i.e. consistent with other logic), so I unresolved it. Is there a reason we'd rather revert on an overflow caused by `* 100` as opposed to using `saturatingMul` there? (cc @geoknee)

(In reality, overflow in any of these computations seems unrealistic because uint256 is so large compared to the magnitude of gasUsed)

geoknee replied:
I don't think there is a reason to do that, and I think we should take the AI's suggestion here.

tynes replied:
This function is going to be called [here](https://github.com/wevm/viem/blob/5031192b959997f6972b862489c11879e8f44353/src/op-stack/actions/estimateTotalFee.ts#L79). It would be a bug in the caller that would result in passing in too large of a number. It could result in creating and signing an invalid tx (not enough balance to pay for fees). Is it better to revert in gas estimation and prevent the invalid tx from being created in the first place? Generally think that this is a product question that we should note how it impacts the user journey and ship whatever is simple. We should develop an opinion in the future on the desired impact on the user journey and make future decisions that align towards that. The failure mode would be apps then depend on the functionality and it changes in the future based on the desired user journey

geoknee replied:
We will circle back to this, tracking here https://github.com/ethereum-optimism/optimism/issues/17888

mds1 commented on packages/contracts-bedrock/test/L2/GasPriceOracle.t.sol:
```diff
@@ -392,4 +399,88 @@ contract GasPriceOracleIsthmus_Test is GasPriceOracle_Test {
+        require(success, "GasPriceOracleJovian_Test: L1Block setup failed");
+
+        assertEq(gasPriceOracle.isIsthmus(), true, "Isthmus should be active before enabling Jovian");
+        assertEq(gasPriceOracle.isJovian(), false, "Jovian starts inactive");
```
typo, since if this assertion failed it means jovian started as active, not inactive?
```suggestion
        assertEq(gasPriceOracle.isJovian(), false, "Jovian starts active");
```

mds1 commented on packages/contracts-bedrock/test/L2/GasPriceOracle.t.sol:
```diff
@@ -392,4 +399,88 @@ contract GasPriceOracleIsthmus_Test is GasPriceOracle_Test {
+        gasPriceOracle.setJovian();
+    }
+
+    /// @dev Tests the transition from Isthmus formula to Jovian formula.
```
All the test cases here seem pretty basic—no fuzz tests, and no tests of edge cases or saturating behavior. We should add some stronger tests for edge case coverage

geoknee replied:
As discussed earlier with @mds1 it is acceptable to add these tests in a followup PR and not block on it.

geoknee replied:
Tracking here https://github.com/ethereum-optimism/optimism/issues/17887

geoknee replied:
Tracking here https://github.com/ethereum-optimism/optimism/issues/17887

mds1 commented on op-service/txinclude/txbudget.go:
```diff
@@ -96,7 +96,9 @@ func (b *TxBudget) AfterIncluded(budgetedCost eth.ETH, tx *IncludedTx) {
 	if receipt.OperatorFeeScalar != nil {
 		// https://github.com/ethereum-optimism/op-geth/blob/6005dd53e1b50fe5a3f59764e3e2056a639eff2f/core/types/rollup_cost.go#L244-L247
-		// Also see: https://specs.optimism.io/protocol/isthmus/exec-engine.html#operator-operatorCost
+		// Also see: https://specs.optimism.io/protocol/isthmus/exec-engine.html#operator-fee
+		// TODO(17817): This currently uses the Isthmus formula (divide by 1e6). Need to update to handle both
+		// Isthmus and Jovian formulas based on fork activation. Jovian formula multiplies by 100 instead.
```
Are we planning to do this TODO in this PR?

geoknee replied:
No, there's a tracking ticket for it so that we don't need to do it here.

leruaa replied:
According to [this comment](https://github.com/ethereum-optimism/optimism/pull/17366#discussion_r2420388887), no

tynes reviewed (approved):
Giving approval but no merge until the dependency is updated

geoknee committed:
Update packages/contracts-bedrock/test/L2/GasPriceOracle.t.sol

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

geoknee committed:
jovian upgrade transactions: reorder deployer addresses

geoknee committed:
just update-op-geth 5c6d276814f2cce1d4dda25ed9e5a3a1c52e59a4

geoknee commented:
/ci authorize [866499f](https://github.com/ethereum-optimism/optimism/pull/17366/commits/866499f22e73ec9c8c94cb63a9490da305e5a128)

geoknee committed:
explicitly disallow user transactions in jovian activation block

geoknee committed:
assert on receipts in operator fee acceptance test

geoknee committed:
op-e2e/actions: remove jovianActivationBlock test case

replace with running the normalTx test case on the jovian fork

geoknee commented:
/ci authorize [b719641](https://github.com/ethereum-optimism/optimism/pull/17366/commits/b71964183b50e3c3bf1d43c97b40d48d0bd73a04)
